### PR TITLE
autoflex/ expand json interface type

### PR DESCRIPTION
**Pull request by johnsonaj**
### Description

First pass at expanding JSON strings into Smithy objects

Uses `jsontypes.Normalized` as the standard for validating the document

### Output from Acceptance Testing

```console
% go test -count=1 ./internal/framework/flex/... -run=TestFlatten\|TestExpand -v

--- PASS: TestFlattenFrameworkStringValueMapLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMapLegacy/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMapLegacy/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMapLegacy/nil_map (0.00s)
--- PASS: TestExpandFrameworkStringSet (0.00s)
    --- PASS: TestExpandFrameworkStringSet/null (0.00s)
    --- PASS: TestExpandFrameworkStringSet/zero_elements (0.00s)
    --- PASS: TestExpandFrameworkStringSet/unknown (0.00s)
    --- PASS: TestExpandFrameworkStringSet/two_elements (0.00s)
    --- PASS: TestExpandFrameworkStringSet/invalid_element_type (0.00s)
--- PASS: TestFlattenFrameworkStringValueMap (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMap/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMap/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueMap/nil_map (0.00s)
--- PASS: TestFlattenSimpleSingleNestedBlock (0.00s)
    --- PASS: TestFlattenSimpleSingleNestedBlock/single_nested_block_pointer (0.00s)
    --- PASS: TestFlattenSimpleSingleNestedBlock/single_nested_block_nil (0.00s)
    --- PASS: TestFlattenSimpleSingleNestedBlock/single_nested_block_value (0.00s)
--- PASS: TestFlattenFrameworkStringMap (0.00s)
    --- PASS: TestFlattenFrameworkStringMap/nil_map (0.00s)
    --- PASS: TestFlattenFrameworkStringMap/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringMap/zero_elements (0.00s)
--- PASS: TestFlattenFrameworkStringValueSetLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSetLegacy/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSetLegacy/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSetLegacy/zero_elements (0.00s)
--- PASS: TestFlattenFrameworkStringValueList (0.00s)
    --- PASS: TestFlattenFrameworkStringValueList/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringValueList/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueList/two_elements (0.00s)
--- PASS: TestFlattenSimpleNestedBlockWithStringEnum (0.00s)
    --- PASS: TestFlattenSimpleNestedBlockWithStringEnum/single_nested_valid_value (0.00s)
    --- PASS: TestFlattenSimpleNestedBlockWithStringEnum/single_nested_empty_value (0.00s)
--- PASS: TestExpandFrameworkStringValueSet (0.00s)
    --- PASS: TestExpandFrameworkStringValueSet/null (0.00s)
    --- PASS: TestExpandFrameworkStringValueSet/zero_elements (0.00s)
    --- PASS: TestExpandFrameworkStringValueSet/invalid_element_type (0.00s)
    --- PASS: TestExpandFrameworkStringValueSet/two_elements (0.00s)
    --- PASS: TestExpandFrameworkStringValueSet/unknown (0.00s)
--- PASS: TestFlattenFrameworkStringValueListLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringValueListLegacy/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueListLegacy/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringValueListLegacy/zero_elements (0.00s)
--- PASS: TestExpandFrameworkStringList (0.00s)
    --- PASS: TestExpandFrameworkStringList/null (0.00s)
    --- PASS: TestExpandFrameworkStringList/invalid_element_type (0.00s)
    --- PASS: TestExpandFrameworkStringList/zero_elements (0.00s)
    --- PASS: TestExpandFrameworkStringList/two_elements (0.00s)
    --- PASS: TestExpandFrameworkStringList/unknown (0.00s)
--- PASS: TestExpandFrameworkStringValueMap (0.00s)
    --- PASS: TestExpandFrameworkStringValueMap/null (0.00s)
    --- PASS: TestExpandFrameworkStringValueMap/zero_elements (0.00s)
    --- PASS: TestExpandFrameworkStringValueMap/two_elements (0.00s)
    --- PASS: TestExpandFrameworkStringValueMap/unknown (0.00s)
    --- PASS: TestExpandFrameworkStringValueMap/invalid_element_type (0.00s)
--- PASS: TestFlattenComplexNestedBlockWithStringEnum (0.00s)
    --- PASS: TestFlattenComplexNestedBlockWithStringEnum/single_nested_zero_value (0.00s)
    --- PASS: TestFlattenComplexNestedBlockWithStringEnum/single_nested_empty_value (0.00s)
    --- PASS: TestFlattenComplexNestedBlockWithStringEnum/single_nested_valid_value (0.00s)
--- PASS: TestFlattenFrameworkStringList (0.00s)
    --- PASS: TestFlattenFrameworkStringList/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringList/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringList/nil_array (0.00s)
--- PASS: TestFlattenComplexSingleNestedBlock (0.00s)
    --- PASS: TestFlattenComplexSingleNestedBlock/single_nested_block_pointer (0.00s)
--- PASS: TestExpandListOfStringEnum (0.00s)
    --- PASS: TestExpandListOfStringEnum/valid_value (0.00s)
    --- PASS: TestExpandListOfStringEnum/null_value (0.00s)
    --- PASS: TestExpandListOfStringEnum/empty_value (0.00s)
--- PASS: TestExpandStringEnum (0.00s)
    --- PASS: TestExpandStringEnum/empty_value (0.00s)
    --- PASS: TestExpandStringEnum/valid_value (0.00s)
--- PASS: TestExpandComplexSingleNestedBlock (0.00s)
    --- PASS: TestExpandComplexSingleNestedBlock/single_nested_block_pointer (0.00s)
--- PASS: TestExpandSimpleNestedBlockWithStringEnum (0.00s)
    --- PASS: TestExpandSimpleNestedBlockWithStringEnum/single_nested_valid_value (0.00s)
    --- PASS: TestExpandSimpleNestedBlockWithStringEnum/single_nested_empty_value (0.00s)
--- PASS: TestExpandFrameworkStringMap (0.00s)
    --- PASS: TestExpandFrameworkStringMap/null (0.00s)
    --- PASS: TestExpandFrameworkStringMap/invalid_element_type (0.00s)
    --- PASS: TestExpandFrameworkStringMap/null_element (0.00s)
    --- PASS: TestExpandFrameworkStringMap/two_elements (0.00s)
    --- PASS: TestExpandFrameworkStringMap/zero_elements (0.00s)
    --- PASS: TestExpandFrameworkStringMap/unknown (0.00s)
--- PASS: TestFlattenFrameworkStringValueSet (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSet/zero_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSet/two_elements (0.00s)
    --- PASS: TestFlattenFrameworkStringValueSet/nil_array (0.00s)
--- PASS: TestExpandComplexNestedBlockWithStringEnum (0.00s)
    --- PASS: TestExpandComplexNestedBlockWithStringEnum/single_nested_valid_value (0.00s)
    --- PASS: TestExpandComplexNestedBlockWithStringEnum/single_nested_empty_value (0.00s)
--- PASS: TestExpandFrameworkStringValueList (0.01s)
    --- PASS: TestExpandFrameworkStringValueList/null (0.00s)
    --- PASS: TestExpandFrameworkStringValueList/zero_elements (0.00s)
    --- PASS: TestExpandFrameworkStringValueList/two_elements (0.00s)
    --- PASS: TestExpandFrameworkStringValueList/invalid_element_type (0.00s)
    --- PASS: TestExpandFrameworkStringValueList/unknown (0.00s)
--- PASS: TestExpandSetOfStringEnum (0.00s)
    --- PASS: TestExpandSetOfStringEnum/valid_value (0.00s)
    --- PASS: TestExpandSetOfStringEnum/null_value (0.00s)
    --- PASS: TestExpandSetOfStringEnum/empty_value (0.00s)
--- PASS: TestFlattenFrameworkStringListLegacy (0.00s)
    --- PASS: TestFlattenFrameworkStringListLegacy/nil_array (0.00s)
    --- PASS: TestFlattenFrameworkStringListLegacy/zero_elements (0.01s)
    --- PASS: TestFlattenFrameworkStringListLegacy/two_elements (0.00s)
--- PASS: TestExpandSimpleSingleNestedBlock (0.00s)
    --- PASS: TestExpandSimpleSingleNestedBlock/single_nested_block_nil (0.00s)
    --- PASS: TestExpandSimpleSingleNestedBlock/single_nested_block_pointer (0.00s)
    --- PASS: TestExpandSimpleSingleNestedBlock/single_nested_block_value (0.00s)
--- PASS: TestFlattenGeneric (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_ptr_both (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_ptr_source (0.00s)
    --- PASS: TestFlattenGeneric/map_block_enum_key (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_set (0.00s)
    --- PASS: TestFlattenGeneric/nested_string_map (0.00s)
    --- PASS: TestFlattenGeneric/map_string (0.00s)
    --- PASS: TestFlattenGeneric/complex_Source_and_complex_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]*struct_and_empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]*struct_and_null_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]*struct_and_empty_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]*struct_and_null_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]struct_and_non-empty_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]struct_and_non-empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]struct_and_empty_struct_Target (0.00s)
    --- PASS: TestFlattenGeneric/empty_[]struct_and_empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]struct_and_null_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]*struct_and_non-empty_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_[]struct_and_null_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/*struct_Source_and_single_set_Target (0.00s)
    --- PASS: TestFlattenGeneric/nil_*struct_Source_and_single_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/*struct_Source_and_single_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/non-empty_[]*struct_and_non-empty_list_Target (0.00s)
    --- PASS: TestFlattenGeneric/map_block_key_list (0.00s)
--- PASS: TestFlatten (0.01s)
    --- PASS: TestFlatten/timestamp_empty (0.00s)
    --- PASS: TestFlatten/timestamp_nil (0.00s)
    --- PASS: TestFlatten/timestamp (0.00s)
    --- PASS: TestFlatten/timestamp_pointer (0.00s)
    --- PASS: TestFlatten/nil_Source_and_Target (0.00s)
    --- PASS: TestFlatten/single_nil_*string_Source_and_single_ARN_Target (0.00s)
    --- PASS: TestFlatten/capitalization_field_names (0.00s)
    --- PASS: TestFlatten/resource_name_prefix (0.00s)
    --- PASS: TestFlatten/slice/map_of_string_types_Source_and_List/Set/Map_of_string_types_Target (0.00s)
    --- PASS: TestFlatten/zero_value_slice/map_of_string_type_Source_and_List/Set/Map_of_string_types_Target (0.00s)
    --- PASS: TestFlatten/slice/map_of_primtive_types_Source_and_List/Set/Map_of_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/zero_value_slice/map_of_primtive_types_Source_and_List/Set/Map_of_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/primtive_types_Source_and_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/zero_value_primtive_types_Source_and_primtive_types_Target (0.00s)
    --- PASS: TestFlatten/single_string_Source_and_single_ARN_Target (0.00s)
    --- PASS: TestFlatten/single_string_Source_and_single_int64_Target (0.00s)
    --- PASS: TestFlatten/single_*string_Source_and_single_ARN_Target (0.00s)
    --- PASS: TestFlatten/single_string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/single_empty_string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/does_not_implement_attr.Value_Target (0.00s)
    --- PASS: TestFlatten/single_*string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/empty_struct_pointer_Source_and_Target (0.00s)
    --- PASS: TestFlatten/single_string_struct_pointer_Source_and_empty_Target (0.00s)
    --- PASS: TestFlatten/json_interface_Source_string_Target (0.00s)
    --- PASS: TestFlatten/non-struct_Target (0.00s)
    --- PASS: TestFlatten/non-struct_Source (0.00s)
    --- PASS: TestFlatten/non-pointer_Target (0.00s)
    --- PASS: TestFlatten/empty_struct_Source_and_Target (0.00s)
    --- PASS: TestFlatten/strange_plurality (0.00s)
    --- PASS: TestFlatten/single_nil_*string_Source_and_single_string_Target (0.00s)
    --- PASS: TestFlatten/plural_ordinary_field_names (0.00s)
    --- PASS: TestFlatten/plural_field_names (0.00s)
--- PASS: TestExpandGeneric (0.01s)
    --- PASS: TestExpandGeneric/single_list_Source_and_*struct_Target (0.00s)
    --- PASS: TestExpandGeneric/map_block_key_set (0.00s)
    --- PASS: TestExpandGeneric/map_block_key_list (0.00s)
    --- PASS: TestExpandGeneric/map_string (0.00s)
    --- PASS: TestExpandGeneric/map_block_key_ptr_both (0.00s)
    --- PASS: TestExpandGeneric/complex_Source_and_complex_Target (0.00s)
    --- PASS: TestExpandGeneric/map_block_enum_key (0.00s)
    --- PASS: TestExpandGeneric/non-empty_list_Source_and_non-empty_[]struct_Target (0.00s)
    --- PASS: TestExpandGeneric/empty_list_Source_and_empty_[]struct_Target (0.00s)
    --- PASS: TestExpandGeneric/map_block_key_ptr_source (0.00s)
    --- PASS: TestExpandGeneric/empty_list_Source_and_empty_[]*struct_Target (0.00s)
    --- PASS: TestExpandGeneric/single_set_Source_and_*struct_Target (0.00s)
    --- PASS: TestExpandGeneric/non-empty_list_Source_and_non-empty_[]*struct_Target (0.00s)
    --- PASS: TestExpandGeneric/non-empty_set_Source_and_non-empty_[]*struct_Target (0.00s)
    --- PASS: TestExpandGeneric/empty_set_Source_and_empty_[]*struct_Target (0.00s)
    --- PASS: TestExpandGeneric/non-empty_list_Source_and_non-empty_[]struct_Target#01 (0.00s)
    --- PASS: TestExpandGeneric/empty_list_Source_and_empty_[]struct_Target#01 (0.00s)
    --- PASS: TestExpandGeneric/non-empty_set_Source_and_non-empty_[]struct_Target (0.00s)
    --- PASS: TestExpandGeneric/nested_string_map (0.00s)
--- PASS: TestExpand (0.02s)
    --- PASS: TestExpand/nil_Source_and_Target (0.00s)
    --- PASS: TestExpand/string_Source_to_interface_Target (0.00s)
    --- PASS: TestExpand/timestamp (0.00s)
    --- PASS: TestExpand/timestamp_pointer (0.00s)
    --- PASS: TestExpand/single_ARN_Source_and_single_*string_Target (0.00s)
    --- PASS: TestExpand/single_ARN_Source_and_single_string_Target (0.00s)
    --- PASS: TestExpand/resource_name_prefix (0.00s)
    --- PASS: TestExpand/capitalization_field_names (0.00s)
    --- PASS: TestExpand/single_string_struct_pointer_Source_and_empty_Target (0.00s)
    --- PASS: TestExpand/List/Set/Map_of_primitive_types_Source_and_slice/map_of_primtive_types_Target (0.00s)
    --- PASS: TestExpand/primtive_types_Source_and_primtive_types_Target (0.00s)
    --- PASS: TestExpand/single_string_Source_and_single_int64_Target (0.00s)
    --- PASS: TestExpand/single_string_Source_and_single_*string_Target (0.00s)
    --- PASS: TestExpand/single_string_Source_and_single_string_Target (0.00s)
    --- PASS: TestExpand/does_not_implement_attr.Value_Source (0.00s)
    --- PASS: TestExpand/types.String_to_string (0.00s)
    --- PASS: TestExpand/empty_struct_pointer_Source_and_Target (0.00s)
    --- PASS: TestExpand/empty_struct_Source_and_Target (0.00s)
    --- PASS: TestExpand/non-struct_Source (0.00s)
    --- PASS: TestExpand/non-struct_Target (0.00s)
    --- PASS: TestExpand/non-pointer_Target (0.00s)
    --- PASS: TestExpand/plural_field_names (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/framework/flex	0.480s
```

### CODE REVIEW

**Comment by johnsonaj** [comment]
```console
% make testacc TESTARGS='-run=TestAccOpenSearchServerlessLifecyclePolicy' PKG=opensearchserverless

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.22.2 test ./internal/service/opensearchserverless/... -v -count 1 -parallel 20  -run=TestAccOpenSearchServerlessLifecyclePolicy -timeout 360m
--- PASS: TestAccOpenSearchServerlessLifecyclePolicy_disappears (12.07s)
--- PASS: TestAccOpenSearchServerlessLifecyclePolicy_basic (13.70s)
--- PASS: TestAccOpenSearchServerlessLifecyclePolicy_update (17.71s)
--- PASS: TestAccOpenSearchServerlessLifecyclePolicyDataSource_basic (23.35s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/opensearchserverless	28.625s
```

**Comment by johnsonaj** [comment]
```console
% go test -count=1 ./internal/framework/types/... -run=TestSmithyJSON -v

--- PASS: TestSmithyJSONValidateAttribute (0.00s)
    --- PASS: TestSmithyJSONValidateAttribute/null_value (0.00s)
    --- PASS: TestSmithyJSONValidateAttribute/unknown_value (0.00s)
    --- PASS: TestSmithyJSONValidateAttribute/invalid_ARN (0.00s)
    --- PASS: TestSmithyJSONValidateAttribute/valid_SmithyJSON (0.00s)
--- PASS: TestSmithyJSONTypeValueFromTerraform (0.00s)
    --- PASS: TestSmithyJSONTypeValueFromTerraform/valid_SmithyJSON (0.00s)
    --- PASS: TestSmithyJSONTypeValueFromTerraform/invalid_ARN (0.00s)
    --- PASS: TestSmithyJSONTypeValueFromTerraform/null_value (0.00s)
    --- PASS: TestSmithyJSONTypeValueFromTerraform/unknown_value (0.00s)
--- PASS: TestSmithyJSONValueInterface (0.00s)
    --- PASS: TestSmithyJSONValueInterface/unknown_value (0.00s)
    --- PASS: TestSmithyJSONValueInterface/valid_SmithyJSON (0.00s)
    --- PASS: TestSmithyJSONValueInterface/null_value (0.00s)
    --- PASS: TestSmithyJSONValueInterface/invalid_SmithyJSON (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/framework/types	0.475s
```

File: internal/framework/types/smithy_json_test.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+		},
+		"unknown value": {
+			val:      tftypes.NewValue(tftypes.String, tftypes.UnknownValue),
+			expected: fwtypes.SmithyJSONUnknown[smithyjson.JSONStringer](),
+		},
+		"valid SmithyJSON": {
+			val:      tftypes.NewValue(tftypes.String, `{"test": "value"}`),                      // lintignore:AWSAT003,AWSAT005
+			expected: fwtypes.SmithyJSONValue[smithyjson.JSONStringer](`{"test": "value"}`, nil), // lintignore:AWSAT003,AWSAT005
+		},
+		"invalid ARN": {
```
Comment (jar-b) [inline internal/framework/types/smithy_json_test.go:38]:
```suggestion
		"invalid SmithyJSON": {
```

File: internal/framework/types/smithy_json_test.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+		"null value": {
+			val: fwtypes.SmithyJSONNull[smithyjson.JSONStringer](),
+		},
+		"unknown value": {
+			val: fwtypes.SmithyJSONUnknown[smithyjson.JSONStringer](),
+		},
+		"valid SmithyJSON": { // lintignore:AWSAT003,AWSAT005
+			val: fwtypes.SmithyJSONValue[smithyjson.JSONStringer](`{"test": "value"}`, nil), // lintignore:AWSAT003,AWSAT005
+		},
+		"invalid ARN": {
```
Comment (jar-b) [inline internal/framework/types/smithy_json_test.go:79]:
```suggestion
		"invalid SmithyJSON": {
```

File: internal/framework/types/smithy_json_test.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		"null value": {
+			val:      tftypes.NewValue(tftypes.String, nil),
+			expected: fwtypes.SmithyJSONNull[smithyjson.JSONStringer](),
+		},
+		"unknown value": {
+			val:      tftypes.NewValue(tftypes.String, tftypes.UnknownValue),
+			expected: fwtypes.SmithyJSONUnknown[smithyjson.JSONStringer](),
+		},
+		"valid SmithyJSON": {
+			val:      tftypes.NewValue(tftypes.String, `{"test": "value"}`),                      // lintignore:AWSAT003,AWSAT005
```
Comment (jar-b) [inline internal/framework/types/smithy_json_test.go:35]: I think the `lintignore` comments can be removed from all of these.

```suggestion
			val:      tftypes.NewValue(tftypes.String, `{"test": "value"}`),
```

**Review by jar-b — APPROVED** [review]
LGTM 🎉 

```console
% go test -count=1 ./internal/framework/...
?       github.com/hashicorp/terraform-provider-aws/internal/framework  [no test files]
ok      github.com/hashicorp/terraform-provider-aws/internal/framework/flex     0.567s
ok      github.com/hashicorp/terraform-provider-aws/internal/framework/types    0.238s
ok      github.com/hashicorp/terraform-provider-aws/internal/framework/validators       0.483s
```

**Pull request merged**